### PR TITLE
FormatChecker.IsFormat parameter as interface{}

### PR DESCRIPTION
**Pull request by ricardomaraschini**
Another patch laying around on my local repo here. As I needed a way to, more than check if a given property is an integer, check if it is a valid integer looking somewhere else(on a database on my case, business logic) I came up with this solution. Maybe it is out of scope for what JSON schema is supposed to support but someone may think it is useful as I think it is.

To achieve that i have to migrate the FormatChecker interface to accept an interface{} parameter instead of only strings(this breaks backward compatibility), this way i can specify something like the following on the json schema:

```json
{
       "type": "integer",
       "format": "ValidUserId"
}
```

I have also added two sample testcases on the test suite. The interface implementation for FormatChecker with this patch becomes:

```go
type ValidUserId struct{}

func (v ValidUserId) IsFormat(input interface{}) bool {
     // do the magic
     return true
}
```

As the encoding/json go library appears to treat numbers as float64 always, the IsFormat on the above sample will receive an float64. I have also added a sample at README file.

Hope that helps and thanks for such great library.

### CODE REVIEW

**Comment by xeipuuv** [comment]
a bit concerned about the "this breaks backward compatibility", what do you guys think ?

**Comment by handrews** [comment]
The latest validation draft still requires format to be a string.  There's been some talk of changes around format, although nothing concrete enough (particularly for non-strings) to work with yet.

It might be better to consider a custom keyword for your use case instead of a custom format?

**Comment by ricardomaraschini** [comment]
> It might be better to consider a custom keyword for your use case instead of a custom format?

Do you mean have something like: 

```json
{
    "UserId": {
        "type": "userid"
    }
}
```

That would be great. For that it might be good to start splitting the current validators we already have(like for integers, numbers, strings, ...) into separated entities. Let me try to explain how i see this.

We implement a way to register validators on the exact same way we already have for formats. Then we convert the validation code we have today and create the default "validators entities", each one associated with its own keyword(integer, string, ...). Once we have that, we make a call available so every user may create custom validators, with any name, through a call. Every validator should implement an interface as follow:

```go
Validator interface {
        IsValid(input interface{}) bool
        Error() string
}
```

On the usercase i have here it would be something like:

```go
type UserIdValidator struct{
    lastError string
}

func (x UserIdValidator) IsValid(userid interface{}) bool {
    // do the magic to check if userid is valid
    // the way i want
    return true
}

func (x UserIdValidator) Error() string {
     // this serves to make it more flexible as
     // i may have different errors when validating
     // the user id
    
     return x.lastError
}

```

Then, on my system i would need a call like:

```go
gojsonschema.Validators.Add("userid", UserIdValidator{})
```

What do you think about that?

**Comment by handrews** [comment]
Two things:
1.  I was misreading the draft when I said that format can only validate strings, so writing a custom format checker that validates other data types is compatible with the spec.

2.  To explain what I meant by custom keyword, I meant custom schema keyword rather than custom value of "type".  The meta-schema enumerates the allowed types, so a custom type value would require a custom meta-schema to override that, and users of that type would not be valid against the regular meta-schema.

On the other hand, the JSON Schema spec explicitly allows unknown keywords, in order to allow extensibility.  So even if you did write a custom meta-schema defining your custom keyword, the resulting schema would still validate against the regular meta-schema (it would just ignore your custom keyword).

But maybe a custom format would actually be better here.

**Comment by ricardomaraschini** [comment]
So, if I understand correctly, to have a custom format checker for other things than "string" is also valid. If that is true i think this patch is still valuable. If my assumption is wrong, please let me help out to get the patch right.

File: README.md [hunk 1]
```diff
@@ -203,7 +203,7 @@ Note in most cases, the err.Details() will be used to generate replacement strin
 ```
 
 ## Formats
-JSON Schema allows for optional "format" property to validate strings against well-known formats. gojsonschema ships with all of the formats defined in the spec that you can use like this:
+JSON Schema allows for optional "format" property to validate strings and numbers against well-known formats. gojsonschema ships with all of the formats defined in the spec that you can use like this:
```
Comment (chrisdostert) [inline README.md]: maybe change "strings and numbers" to "instances" (to be consistent w/ terminology spec uses)

**Comment by chrisdostert** [comment]
@ricardomaraschini is it possible to resolve conflicts? @xeipuuv does this seem like a reasonable change? Per my reading of the spec I think this seems like a nice addition  : )

File: README.md (hunk 1 above)
Comment (handrews) [reply README.md]: I agree that "instances" is preferable, although it should match the actual support.  if the next step is just to get numbers as well as strings, that would be good (there are definitely numeric format suggestions, but while I can make up some formats for other types none are actually motivated by use cases :-)

**Pull request closed without merging**